lib/std/crypto/rsa.zig

```diff
@@ -0,0 +1,731 @@
+                    // > satisfying e * d == 1 (mod \lambda(n)),
+                    if (!d.isOdd()) return error.InvalidExponent;
+                    if (d.v.compare(n.v) != .lt) return error.InvalidExponent;
+                }
+
+                return .{ .private_exponent = d };
+            }
+        };
+
+        pub const KeyPair = struct {
+            public: PublicKey,
+            secret: SecretKey,
+
+            pub fn fromDer(parser: *der.Parser) !KeyPair {
+                // We're just interested in the first few fields which don't vary by version
+                _ = try parser.expectSequence();
+                const version = try parser.expectPrimitive(.integer);
+                _ = version;
+
+                const mod = try parseModulus(parser);
+                const pub_exp = try parseInteger(parser);
+                const sec_exp = try parseInteger(parser);
+                // Skip unused private key fields
+
+                const public = try PublicKey.fromBytes(mod, pub_exp);
+                const secret = try SecretKey.fromBytes(public.modulus, sec_exp);
+
+                // TODO: make this work (FeUnit multiplication?)
+                if (std.debug.runtime_safety) {
+                    // const prime1 = try parseInteger(&parser);
+                    // const prime2 = try parseInteger(&parser);
+                    // const p = try Fe.fromBytes(public.modulus, prime1, .big);
+                    // const q = try Fe.fromBytes(public.modulus, prime2, .big);
+
+                    // // check that n = p * q
+                    // const n_expected = p.v.mul(q.v); // can't mul these :(
+                    // if (n_expected.v.compare(public.n.v) != .eq) return error.KeyMismatch;
+
+                    // // check that d * e is one mod p-1 and mod q-1. Note d and e were bound
+                    // const de = secret.d.mul(public.e); // can't mul these :(
+                    // const one = public.n.one();
+
+                    // if (public.n.mul(de, p.sub(one)).compare(one) != .eq) return error.KeyMismatch;
+                    // if (public.n.mul(de, q.sub(one)).compare(one) != .eq) return error.KeyMismatch;
+                }
+
+                return .{ .public = public, .secret = secret };
+            }
+
+            /// Decrypt a short message using RSAES-PKCS1-v1_5.
+            pub fn decrypt(kp: KeyPair, ciphertext: [modulus_len]u8, out: *[modulus_len]u8) ![]u8 {
+                const m = Fe.fromBytes(kp.public.modulus, &ciphertext, .big) catch return error.MessageTooLong;
+                const e = kp.public.modulus.pow(m, kp.secret.private_exponent) catch unreachable;
+                try e.toBytes(out, .big);
+
+                const msg_start = std.mem.lastIndexOfScalar(u8, out, 0);
+                // Care shall be taken to ensure that an opponent cannot
+                // distinguish these error conditions, whether by error
+                // message or timing.
```

Comment:
For that alone, `rsa` cannot be public, or that will be an immediate CVE. It's been an recurring issue in all crypto libraries, the Rust RSA crate is in an embarrassing situation that they are unable to fix; we should do better before declaring ours public. And at least follow https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-rsa-guidance-00

Reply:
Ah, I thought putting the checks in the same branch would be enough:
```zig
if (out[0] != 0 or out[1] != 2 or ps_len < 8) return error.DecryptionError;
```

I'll check the assembly, read the link, and see why I'm wrong.

Reply:
So does `std.mem.lastIndexOfScalar()`. So does `out[i + 1 ..]`.  PKCS#1.5 is nasty.

Reply:
Let's catch em all! I wish most certificate chains didn't include these :(

I'll also update it to be deprecated.